src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,124 @@
-package core.basesyntax;
-
-public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
-
-    }
-
-    @Override
-    public void add(T value, int index) {
-
-    }
-
-    @Override
-    public void addAll(List<T> list) {
-
-    }
-
-    @Override
-    public T get(int index) {
-        return null;
-    }
-
-    @Override
-    public void set(T value, int index) {
-
-    }
-
-    @Override
-    public T remove(int index) {
-        return null;
-    }
-
-    @Override
-    public T remove(T element) {
-        return null;
-    }
-
-    @Override
-    public int size() {
-        return 0;
-    }
-
-    @Override
-    public boolean isEmpty() {
-        return false;
-    }
-}
+package core.basesyntax;
+
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
+public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private T[] grow() {
+        return elementData = Arrays.copyOf(elementData, newCapacity());
+    }
+
+    private int newCapacity() {
+        return elementData.length + (elementData.length >> 1);
+    }
+
+    @Override
+    public void add(T value) {
+        if (size == elementData.length) {
+            grow();
+        }
+        elementData[size] = value;
+        size++;
+    }
+
+    @Override
+    public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: "
+                    + index + " size: " + size);
+        }
+        T[] elementData;
+        if (size == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                size - index);
+        elementData[index] = value;
+        size++;
+    }
+
+    @Override
+    public void addAll(List<T> list) {
+        T[] currentList = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            currentList[i] = list.get(i);
+        }
+        if ((elementData.length - size) < currentList.length) {
+            elementData = grow();
+        }
+        System.arraycopy(currentList, 0, elementData, size, currentList.length);
+        size = size + currentList.length;
+    }
+
+    @Override
+    public T get(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bound for length " + size);
+        }

```

Comment:
you use the same block of code in several places, let's create a separate method and move this logic there